pkg/policy: remove ReverseRules from Consumable

ianvernon opened the pull request:
If connection-tracking was not enabled for an endpoint, in order to ensure that
traffic between endpoints managed by Cilium would flow, if a policy allowed
ingress to endpoint A from endpoint B, then B's policy maps were also updated
to allow ingress to itself from A as well. However, this only worked if both
endpoints were local to the same node, as the sharing of this information is
done on a per-daemon level; security identities are not resolved on each node
in the cluster. As such, remove ReverseRules, which was used to store such data,
from the Consumable structure.

There is a special case involving the reserved-identities, whose corresponding
Consumables do need to be updated if an endpoint allows traffic to them. 

Signed-off by: Ian Vernon <ian@cilium.io>

Fixes: #2795

Commit:
pkg/policy: remove ReverseRules from Consumable

If connection-tracking was not enabled for an endpoint, in order to ensure that
traffic between endpoints managed by Cilium would flow, if a policy allowed
ingress to endpoint A from endpoint B, then B's policy maps were also updated
to allow ingress to itself from A as well. However, this only worked if both
endpoints were local to the same node, as the sharing of this information is
done on a per-daemon level; security identities are not resolved on each node
in the cluster. As such, remove ReverseRules, which was used to store such data,
from the Consumable structure.

Signed-off by: Ian Vernon <ian@cilium.io>

joestringer reviewed (dismissed):
LGTM, less cases to worry about.

CT can still be disabled with this patch, but users must disable CT for all endpoints that intend to still be able to talk to one another. This should only be used if you know what you're doing.

Commit:
pkg/policy: finish incomplete documentation of Init

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/endpoint: specify that Consumable mutex must be held during regenerateConsumable

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: add IsReservedIdentity function

Add function which checks whether the specified NumericIdentity corresponds to a reserved identity.

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
pkg/policy: explicitly update reserved Consumables

The removal of ReverseRules necessitates explicit updating of Consumables which
correspond to the reserved identities, because Consumables are only regenerated
for endpoints. Since the Consumables representing reserved identities are not
related to endpoints, they have to be updated accordingly whenever a Consumable
corresponding to an endpoint is updated (that is, whenever an endpoint allows
a reserved identity, the Consumable representing the reserved identity must be
updated to allow traffic from the endpoint's Consumable).

Signed-off by: Ian Vernon <ian@cilium.io>

Commit:
test: update test to work with removal of ReverseRules

* Import a policy that allows traffic in both directions when Conntrack is
disabled for an endpoint.
* Fix some improper log messages
* Dump output of `cilium policy get` during connectivity test.

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on pkg/policy/consumer.go:
```diff
@@ -205,44 +179,32 @@ func (c *Consumable) AllowIngressIdentityLocked(cache *ConsumableCache, id Numer
+					reservedConsumable.AllowIngressIdentityLocked(cache, c.ID)
+					reservedConsumable.Mutex.Unlock()
+				} else {
+					reservedConsumable.AllowIngressIdentityLocked(cache, c.ID)
```
Why is the logic exactly the same on both branches?

ianvernon replied:
It is not the same; one acquires and releases the lock, and one does not. If a Consumable can allow itself (which it can IIRC), then we should not call the lock for it because it will already be locked by the caller of `AllowIngressIdentityLocked`.

aanm replied:
Please add this comment in the code

ianvernon replied:
Done.

ianvernon replied:
Sorry, now it is done - didn't push to origin for the branch before I commented.

aanm commented on pkg/policy/consumer.go:
```diff
@@ -253,11 +215,27 @@ func (c *Consumable) RemoveIngressIdentityLocked(id NumericIdentity) {
+					reservedConsumable.RemoveIngressIdentityLocked(c.ID)
+					reservedConsumable.Mutex.Unlock()
+				} else {
+					reservedConsumable.RemoveIngressIdentityLocked(c.ID)
```
same here?

ianvernon replied:
Same as above reply in `AllowIngressIdentityLocked`.

aanm commented on pkg/policy/identity.go:
```diff
@@ -72,6 +72,16 @@ func (id NumericIdentity) String() string {
 
+// IsReservedIdentity returns whether id is one of the special reserved identities.
+func (id NumericIdentity) IsReservedIdentity() bool {
+	for reservedIdentity := range ReservedIdentityNames {
```
why not `_, e := ReservedIdentityNames[id]; return e`?

ianvernon replied:
Heh, good idea. Will fix in next upload.

ianvernon replied:
Fixed.

jrajahalme reviewed (approved):
Formal approval for proxy, did not read any code :-)

Commit:
pkg/policy: specify when lock is acquired for reserved identity Consumables

Signed-off by: Ian Vernon <ian@cilium.io>

aanm commented on test/runtime/connectivity.go:
```diff
@@ -320,9 +320,13 @@ var _ = Describe("RuntimeValidatedConntrackTest", func() {
 
+		By("Showing policies imported to Cilium")
+		res := vm.PolicyGetAll()
+		fmt.Printf("%s\n", res.CombineOutput())
```
Just to be sure, is this a leftover?

ianvernon replied:
I left this because it clearly displays what policies are imported in Cilium to console without wrapping it in anything else - I can remove it if you like, though.
